### PR TITLE
Feat rust newtypes members

### PR DESCRIPTION
**Pull request by MikaelMayer**
### Fixes to the Rust compiler
- Support for newtypes wrapping other newtypes
- Full conversions between newtypes and primitive types
- Support for methods on newtypes
- It also adds an attribute `{:rust_erased true/false}` for newtypes to override the default erasure setting for newtypes, which is:
  * Newtypes without members: Erased (so that we can see Rust's native types like u8, u32, etc.)
  * Newtypes with members: Not erased (to improve legibility of the code in Rust)
- Support for general newtypes on booleans
- Support in the generated Rust AST for metadata on types, to know if a type supports Copy or now. Previously it had to be known statically for primitive, but now with newtypes we need to embed this information into the tree.
- Support for wrapping operations for bitvectors and newtypes based on bitvectors

### Fixes for new resolver
- I fixed in the new resolver the case of ITEExpr where it was overriding the PreType of the test instead of constraining it.

### Refactoring
- WrBuffer() makes it easy to handle buffers

### How has this been tested?
- Modified the test comp/rust/newtypes.dfy to support all the new cases, including use of the attribute `{:rust_erased}` and various cases
- Added test case newtypesrefresh.dfy for the newtypes on booleans
- Had to fix the code to make existing tests to pass

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/Backends/Dafny/AST.dfy [hunk 1]
```diff
@@ -130,10 +130,30 @@ module {:extern "DAST"} DAST {
...
+    | U64(overflow: bool)
+    | I64(overflow: bool)
+    | U128(overflow: bool)
+    | I128(overflow: bool)
+    | USIZE  // For whatever target considers that native arrays can be indexed with
+    | BigInt
+    | Bool
     | NoRange
+  {
+    predicate canOverflow() {
```
Comment (keyboardDrummer) [inline Source/DafnyCore/Backends/Dafny/AST.dfy:150]: Why lowercase?

File: Source/DafnyCore/Backends/Dafny/AST.dfy [hunk 2]
```diff
@@ -130,10 +130,30 @@ module {:extern "DAST"} DAST {
...
+    | I8(overflow: bool)
+    | U16(overflow: bool)
+    | I16(overflow: bool)
+    | U32(overflow: bool)
+    | I32(overflow: bool)
+    | U64(overflow: bool)
+    | I64(overflow: bool)
+    | U128(overflow: bool)
+    | I128(overflow: bool)
+    | USIZE  // For whatever target considers that native arrays can be indexed with
```
Comment (keyboardDrummer) [inline Source/DafnyCore/Backends/Dafny/AST.dfy:145]: `USIZE` => `NativeArrayIndex` ?

**Review by keyboardDrummer — APPROVED** [review]
To be honest I don't know this codebase well enough to have much of an opinion about it, but the changes look fine and apparently they improve Rust support.

File: Source/DafnyCore/Backends/Dafny/AST.dfy [hunk 3]
```diff
@@ -130,10 +130,30 @@ module {:extern "DAST"} DAST {
 
   datatype Primitive = Int | Real | String | Bool | Char | Native
 
-  // USIZE is for whatever target considers that native arrays can be indexed with
+
   datatype NewtypeRange =
-    | U8 | I8 | U16 | I16 | U32 | I32 | U64 | I64 | U128 | I128 | USIZE | BigInt
```
Comment (keyboardDrummer) [inline Source/DafnyCore/Backends/Dafny/AST.dfy:135]: Given that this is in a file AST, does it conceptually mirror the Dafny C# AST?

Comment (MikaelMayer) [reply Source/DafnyCore/Backends/Dafny/AST.dfy:135]: Almost, it's a desugared version of the C# AST, the intermediate representation. For example, comprehension expressions have disappeared, no distinction between functions and methods, and subset types become type synonyms.

File: Source/DafnyCore/Backends/Dafny/AST.dfy (hunk 1 above)
Comment (MikaelMayer) [reply Source/DafnyCore/Backends/Dafny/AST.dfy:150]: I should be more careful about casing, thanks for reporting.

File: Source/DafnyCore/Backends/Dafny/AST.dfy (hunk 2 above)
Comment (MikaelMayer) [reply Source/DafnyCore/Backends/Dafny/AST.dfy:145]: Excellent suggestion, that way I can remove the comment

**Review by keyboardDrummer — APPROVED** [review]

**Pull request merged**